### PR TITLE
IGNITE-13366 Maintenance Mode for corrupted PDS case

### PR DESCRIPTION
**Pull request by sergey-chugunov-1985**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceRegistry.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+     * or null if no tasks were registered for this ID.
+     */
+    public @Nullable MaintenanceTask registerMaintenanceTask(MaintenanceTask task) throws IgniteCheckedException;
+
+    /**
+     * Deletes {@link MaintenanceTask} of given ID from maintenance registry.
+     *
+     * @param mntcId
+     */
+    public void unregisterMaintenanceTask(UUID mntcId);
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceRegistry.java:96]: `maitenanceId`

File: modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceRegistry.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+     *
+     * {@link MaintenanceTask} becomes active when node enters Maintenance Mode and doesn't resolve the task
+     * during maintenance prepare phase.
+     *
+     * @return {@link MaintenanceTask} object for given maintenance ID or null if no maintenance task was found.
+     */
+    @Nullable public MaintenanceTask activeMaintenanceTask(UUID maitenanceId);
+
+    /**
+     * @param id UUID of {@link MaintenanceTask} this callback is registered for.
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceRegistry.java:110]: Missing description

File: modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceRegistry.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+     * @param cb {@link MaintenanceWorkflowCallback} interface used by MaintenanceRegistry to execute
+     *                                              maintenance steps by workflow.
+     */
+    public void registerWorkflowCallback(@NotNull UUID id, @NotNull MaintenanceWorkflowCallback cb);
+
+    /**
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceRegistry.java:116]: Empty javadoc

File: modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceTask.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+package org.apache.ignite.maintenance;
+
+import java.util.UUID;
+
+import org.apache.ignite.lang.IgniteExperimental;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Represents request to handle maintenance situation stored on disk.
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceTask.java:27]: Not clear - what is stored on disk? Do we need to mention that at all?

File: modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceTask.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+     * @param description Mandatory description of maintenance situation.
+     * @param params Optional parameters that may be needed to perform maintenance actions.
+     */
+    public MaintenanceTask(UUID id, String description, String params) {
+        this.id = id;
+        this.description = description;
+        this.params = params;
+    }
+
+    /** */
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceTask.java:68]: Empty javadoc here and below

File: modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceWorkflowCallback.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    /**
+     * Called by {@link MaintenanceRegistry} and enables it to check if maintenance is still needed
+     * for component that provided this callback.
+     *
+     * User may fix maintenance situation by hand when node was down thus before going to maintenance mode
+     * we should be able to check if it is still necessary.
+     *
+     * @return {@code True} if maintenance is still needed for the component.
+     */
+    public boolean proceedWithMaintenance();
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceWorkflowCallback.java:48]: `proceedWithMaintenance` means "start maintenance right now". But we want to check whether maintenance should be started, so a better name would be `shouldProceedWithMaintenance`

File: modules/core/src/main/java/org/apache/ignite/maintenance/package-info.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * <!-- Package description. -->
```
Comment (ptupitsyn) [inline modules/core/src/main/java/org/apache/ignite/maintenance/package-info.java:19]: Please remove the placeholder

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceRegistry.java (hunk 1 above)
Comment (sergey-chugunov-1985) [reply modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceRegistry.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceRegistry.java (hunk 2 above)
Comment (sergey-chugunov-1985) [reply modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceRegistry.java]: fixed javadoc here and in other places in this class

File: modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceTask.java (hunk 4 above)
Comment (sergey-chugunov-1985) [reply modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceTask.java]: I removed it from javadoc but we may want to cover this question in documentation. There is an intention to make file with Maintenance Tasks readable by user so he/she may read information from it and even delete it completely if maintenance situation is fixed by hands.

File: modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceWorkflowCallback.java (hunk 6 above)
Comment (sergey-chugunov-1985) [reply modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceWorkflowCallback.java]: agree, renamed as you suggested

File: modules/core/src/main/java/org/apache/ignite/maintenance/package-info.java (hunk 7 above)
Comment (sergey-chugunov-1985) [reply modules/core/src/main/java/org/apache/ignite/maintenance/package-info.java]: fixed

**Pull request closed without merging**